### ISSUE TITLE
Assigned type (Class) doesn't match the function argument's type (Class)

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#34161]
Godot 3.2 beta2

I am getting a strange error with typed GDScript, where it finds incompatibility between argument type and parameter type... while they are actually the same.
```
SCRIPT ERROR: GDScript::reload: Parse Error: At "update_episode_data_from_file()" call, argument 1. Assigned type (Project) doesn't match the function argument's type (Project).
          At: res://main.gd:294
```

I have a `script_data.gd` file with a `Project` class inside.

I also have a `script_parser.gd` file which const-imports the previous one, in order to define the following function:
```gdscript
const ScriptData = preload("./script_data.gd")
   ...
static func update_episode_data_from_file(project: ScriptData.Project, path: String) -> Array:
   ...
```

 Then I const-imported both scripts inside `main.gd`, in which I have the following:
```gdscript
const ScriptData = preload("./script_data.gd")
const ScriptParser = preload("./script_parser.gd")
    ...
var _project : ScriptData.Project = null
    ...
func _open_project(fpath: String):
    ...
	var dir := fpath.get_base_dir()
	for ep_file_rpath in data.episode_files:
		var path := dir.plus_file(ep_file_rpath) as String
		ScriptParser.update_episode_data_from_file(_project, path)
```

This code works fine, but recently I've been getting this typing error in the editor, and it obscures any error happening elsewhere after this  line...
So far the only thing that fixes it is to close the editor and reopen it. Then a few hours later, it reappeared again.
I tried to repro in a simpler project, but could not find a pattern yet...

**Comment by katoneko** [comment]
Can confirm. Had it in 3.1.1, updated to 3.1.2 in hopes it would go away, it didn't and is present in 3.1.2 as well.

I have script A that I preload into a constant in script B and from there I call a function that's defined in A. It returns a class instance and puts it into a variable in B whose type matches the func's return type, and that's where the errors shows up (it says the return type and the var's type don't match although they clearly do).

In my case, though, I remove a symbol from A, type it back, save the script, and the error goes away until something else happens that brings it back that I haven't figured out yet.

**Comment by Faless** [comment]
I'm experiencing a similar problem when using `class_name` (which should `preload` under the hood IIRC).

This problem seems to be specifically tied to inner classes.

Attached the minimum reproduction project: [proj_gdstatic.zip](https://github.com/godotengine/godot/files/3968820/proj_gdstatic.zip)

Or simply create those 2 files:

```gdscript
# ClassOne.gd
extends Reference
class_name ClassOne

class SubClass:
	
	func _init():
		return

static func create_subclass():
	return SubClass.new()

static func create_subclass_with_type() -> SubClass:
	return create_subclass()
```

```gdscript
# Control.gd (main scene)
extends Control

var my_var : ClassOne.SubClass

func _ready():
	assert(ClassOne.SubClass.new() is ClassOne.SubClass)
	assert(ClassOne.create_subclass() is ClassOne.SubClass)
	my_var = ClassOne.create_subclass() # Okay
	my_var = ClassOne.SubClass.new() # Error!
	my_var = ClassOne.create_subclass_with_type() # Error!
```

**Comment by Faless** [comment]
Maybe related to #30557

**Comment by katoneko** [comment]
Although it's marked as fixed, I'm getting the error again on 3.2. Should I make a new issue?
![image](https://user-images.githubusercontent.com/48890128/75569405-d4242200-5a76-11ea-800a-338fed5fce60.png)
![image](https://user-images.githubusercontent.com/48890128/75569418-db4b3000-5a76-11ea-9e89-b4f51d09d424.png)

**Comment by vnen** [comment]
@katoneko please provide a minimal reproduction project.

**Comment by katoneko** [comment]
I haven't dug into the exact scenario under which the bug reveals itself yet. As soon as I can output a solid MPR, I'll upload it here!

**Comment by katoneko** [comment]
@vnen 
MRP: [class_bug_test.zip](https://github.com/godotengine/godot/files/4332997/class_bug_test.zip)

1. Open the project, the error shouldn't be there yet.
2. Open and edit `inner_class_here.gd` in some way. Erase a character, add another class, write a func, anything goes.
3. Save the changes.
4. Open `use_inner_class.gd`. Depending on your configured Idle Parse Delay, after the script is re-parsed, the error `The assigned value's type (InnerClass) doesn't match the variable's type (InnerClass)` should appear without intervention.
5. Navigate to `return_inner_func.gd`. Edit it in the same way described in step 2.
6. Save the changes.
7. Go back to `use_inner_class.gd`. After the script is re-parsed, the error should go away on its own.

Go to step 2 if you want to repeat the cycle.

**Comment by TheSHEEEP** [comment]
Can confirm this is still a problem in 3.2.2 stable.
Though even editing like described here doesn't "solve" the problem for me - no matter what I do, even at runtime GDScript claims a type is not compatible with itself - which is a bit silly ;)

So far the only workaround I found is to not assign the inner class type, which of course defeats the purpose of being typed in the first place.

**Comment by TheSHEEEP** [comment]
Another workaround that does the job for me:

If this produces the error:
```
func doSomethingFunction(argument :MyInnerClass)
    var instance :MyInnerClass = argument # Doesn't match variable's type error

```
This solves it for me, via using a temporary typeless variable:
```
func doSomethingFunction(argument :MyInnerClass)
    var temp = argument
    var instance :MyInnerClass = temp # No error, works fine
```

**Comment by katoneko** [comment]
Just tested my own MRP and the issue's still present in 3.2.3, just in case. But in my case though, editing does make the issue go away temporarily and doesn't affect runtime.

**Comment by khamarr3524** [comment]
This seems to still be an issue on 3.3.2

**Comment by nene** [comment]
Still an issue in 3.4.2

I don't use inner classes, but I do use `class_name`, which seems to cause this as well. It happens at seemingly random places. Quitting and reopening the project often makes the error disappear.

At other times I've had problems with inheritance. For example when class `Child` extends `Parent` and there's a function that expects `Parent`, but I'm passing in `Child`, which should be fine, but Godot complains with `Assigned type (Child) doesn't match the function argument's type (Parent)`.

**Comment by WarrenMarshall** [comment]
"I don't use inner classes, but I do use class_name, which seems to cause this as well. It happens at seemingly random places."

Confirming. It just pops up for seemingly no reason in places. It doesn't cause runtime problems so I've learned to just ignore it and live with it, but it's annoying.

**Comment by gabstv** [comment]
![image](https://user-images.githubusercontent.com/1563110/183303920-913315c5-578c-433c-bcab-4f6ea42f0f45.png)
It still happens on the 3.5 language server. Like @WarrenMarshall said, works fine during runtime, it's just an annoying false positive.

**Comment by Azagwen** [comment]
I believe this still happens in 4.1, as I just ran into this absurd error:
![image](https://github.com/godotengine/godot/assets/30528331/ca9c76df-63b4-4356-885f-8b50d6bfe5d3)
Godot seemingly thinks the Enum I'm giving to `new()` is different from the one in my class

**Comment by AThousandShips** [comment]
@Azagwen are you sure this is the same error? `Orientation` is a global enum, does this happen if you name it something different?

**Comment by Azagwen** [comment]
@AThousandShips 
The engine has given me no indication that `Orientation` was a global enum, it would even direct me to my own Enum when I would CTRL click on it 🤔

I guess that explains it tho, and it also makes the error I was getting incorrect for my case, even tho it is similar to the one this issue is reporting

**Comment by AThousandShips** [comment]
That seems to be a bug in the shadowing warnings, please open a new issue for this, as the code works fine (minus the fact that you can't make the const variables) once you fix that part

**Comment by Azagwen** [comment]
> @Azagwen are you sure this is the same error? `Orientation` is a global enum, does this happen if you name it something different?
> 
> Edit: Confirmed, this is caused by the fact that `Orientation` is already defined, this should probably instead be an error when defining an Enum that shadows a global name

Are you sure `Orientation` is a global enum? or am I mis-undersstanding?
I can't seem to access it at all from my project, which if it is a global enum explains why I didn't notice it.

**Comment by AThousandShips** [comment]
Yes: [here](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#enum-globalscope-orientation)

**Comment by Azagwen** [comment]
> Yes: [here](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#enum-globalscope-orientation), not the one in `PlaneMesh`

I see, then it being unaccessible is also a bug I guess 🤔

**Comment by Azagwen** [comment]
> I believe this still happens in 4.1, as I just ran into this absurd error: ![image](https://user-images.githubusercontent.com/30528331/266657409-ca9c76df-63b4-4356-885f-8b50d6bfe5d3.png) Godot seemingly thinks the Enum I'm giving to `new()` is different from the one in my class

issue created for this at #81460 Hope I provided enough info

**Comment by renolc** [comment]
Just hit this in 3.5.2.stable. No inner classes, no inheritance, nothing like that. The most complex thing I'm doing is defining `class_name` and that's it. I have a World scene that instantiates and sets up various other entities, usually kickstarted via a signal. I started seeing errors that `(Thing) does not match the expected type of (Thing)`. If I remove the type annotations in my method, it seems to go away, but I like having the auto complete. If I edit the `Thing` script in some trivial way (add a newline for example), then the World script doesn't complain anymore. 🤷‍♀️
